### PR TITLE
Minimal readme file

### LINKED ISSUE #2: Create minimal README.md file
Create a README file with some basic information and instructions to run the project's main tasks, such as:

- How to Install development environment.
- Dataset preprocessing.
- Model training.
- Model evaluation.

### PR DESCRIPTION
**Pull request by joaoherrera**
# What is new
- Minimal version of README.md
- Small adjustments in the `build_container.sh` script (main goal was to set some arguments and document them in the README file)
- [NOT TESTED] Create arguments for training (so they can also be documented in the README file)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+```shell
+cd docker
+bash build_container.sh --data-path <DATA_PATH>
+```
+Where:
+```
+DATA_PATH is the path to a folder containing the dataset (images).
+```
+
+## Dataset Preprocessing
```
Comment (brunodrugowick) [inline README.md:25]:
```suggestion
## Dataset Preprocessing

```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+There are a few scripts available to preprocess data. They can be found in the [standalone](src/standalone/) directory.
```
Comment (brunodrugowick) [inline README.md]:
```suggestion
There are a few scripts available to preprocess data. They can be found in the [src/standalone](src/standalone/) directory.
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+python dataset_patches_generation.py --input-directory <INPUT_DIRECTORY> --input-annotations <INPUT_ANNOTATIONS> --output-directory <OUTPUT_DIRECTORY>
+```
+Where
+```
+INPUT_DIRECTORY: Path to the directory containing images with annotations.
+INPUT_ANNOTATIONS: Path to the images' annotation file.
+OUTPUT_DIRECTORY: Path to the output directory.
+```
+
+## Model Training & Evaluation
```
Comment (brunodrugowick) [inline README.md:53]:
```suggestion
## Model Training & Evaluation

```

File: README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# AI Palatal Rugoscopy
+
+Instance segmentation of palatal rugae patterns using Deep Learning.
+
+
```
Comment (brunodrugowick) [inline README.md]:
```suggestion
```

File: README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+## Development environment
+By default, we use a [Docker](https://www.docker.com/) container as development environment. \
```
Comment (brunodrugowick) [inline README.md]:
```suggestion

By default, we use a [Docker](https://www.docker.com/) container as the development environment.
```

File: README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+Before building the image, please ensure you have the following resources available:
+
+- GPU with an updated driver version
+- nvidia-container-tookit (allows running GPU-accelerated containers)
```
Comment (brunodrugowick) [inline README.md]:
```suggestion
- NVIDIA GPU with an updated driver version
- `nvidia-container-tookit` (allows running GPU-accelerated containers)
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+The script [build-container.sh](docker/build-container.sh) automatically builds the image and creates the container. To execute this script, run the following command in a bash terminal:
```
Comment (brunodrugowick) [inline README.md]:
```suggestion
The script [`docker/build-container.sh`](docker/build-container.sh) automatically builds the image and creates the container. To execute this script, run the following command in a bash terminal:
```

File: README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+```shell
+cd docker
+bash build_container.sh --data-path <DATA_PATH>
+```
+Where:
```
Comment (brunodrugowick) [inline README.md:20]:
```suggestion

Where:

```

File: README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+## Dataset Preprocessing
+There are a few scripts available to preprocess data. They can be found in the [standalone](src/standalone/) directory.
+
+- `dataset_split.py`: Splits a dataset into `train`, `test` and `validation` sets. Usage: 
+```shell
+python dataset_split.py --annotations-path <ANNOTATION_FILE> --output-path <OUTPUT_DIRECTORY> --split <TRAIN_SIZE> <VALIDATION_SIZE> [<TEST_SIZE>]
+```
+Where
```
Comment (brunodrugowick) [inline README.md]:
```suggestion

Where

```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+DATA_PATH is the path to a folder containing the dataset (images).
+```
+
+## Dataset Preprocessing
+There are a few scripts available to preprocess data. They can be found in the [standalone](src/standalone/) directory.
+
+- `dataset_split.py`: Splits a dataset into `train`, `test` and `validation` sets. Usage: 
+```shell
+python dataset_split.py --annotations-path <ANNOTATION_FILE> --output-path <OUTPUT_DIRECTORY> --split <TRAIN_SIZE> <VALIDATION_SIZE> [<TEST_SIZE>]
+```
```
Comment (brunodrugowick) [inline README.md]: You could indent with 2 spaces if you want this code block to align with the text in the bullet.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+ANNOTATION_FILE: Path to annotation file.
+OUTPUT_DIRECTORY: Path to the output file.
+TRAIN_SIZE: Percentage of samples to be allocated in the train subset.
+VALIDATION_SIZE: Percentage of samples to be allocated in the validation subset.
+TEST_SIZE (Optional): Percentage of samples to be allocated in the test subset.
+```
+
+- `dataset_patches_generation.py`: Break images into patches. Usage:
+
```
Comment (brunodrugowick) [inline README.md:46]: You're not very consistent with your line breaks, it appears? 😆 
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+```shell
+python dataset_patches_generation.py --input-directory <INPUT_DIRECTORY> --input-annotations <INPUT_ANNOTATIONS> --output-directory <OUTPUT_DIRECTORY>
+```
```
Comment (brunodrugowick) [inline README.md]: ditto (the indentation that I mentioned before)

File: README.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+Where
```
Comment (brunodrugowick) [inline README.md]:
```suggestion

Where

```

File: README.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+```
+INPUT_DIRECTORY: Path to the directory containing images with annotations.
+INPUT_ANNOTATIONS: Path to the images' annotation file.
+OUTPUT_DIRECTORY: Path to the output directory.
+```
+
+## Model Training & Evaluation
+Training a Deep Learning model for a given computer vision task is possible at
```
Comment (brunodrugowick) [inline README.md]:
```suggestion
Training a Deep Learning model for a given computer vision task is taken care of by
```

File: README.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+```shell
+python train.py 
+    --training-images <TRAINING_IMAGES_DIRECTORY> 
+    --training-annotations <TRAINING_ANNOTATIONS_PATH>
+    --validation-images <VALIDATION_IMAGES_DIRECTORY>
+    --validation-annotations <VALIDATION_ANNOTATIONS_PATH>
+    --output-path <OUTPUT_DIRECTORY>
+    --batch-size <BATCH_SIZE>
+    --epochs <EPOCHS>
```
Comment (brunodrugowick) [inline README.md]: Don't you need `\` at the end of each line?

```suggestion
python train.py \
    --training-images <TRAINING_IMAGES_DIRECTORY> \ 
    --training-annotations <TRAINING_ANNOTATIONS_PATH> \
    --validation-images <VALIDATION_IMAGES_DIRECTORY> \
    --validation-annotations <VALIDATION_ANNOTATIONS_PATH> \
    --output-path <OUTPUT_DIRECTORY> \
    --batch-size <BATCH_SIZE> \
    --epochs <EPOCHS>
```

File: README.md [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+```
+Where:
```
Comment (brunodrugowick) [inline README.md:69]:
```suggestion

Where:

```

File: README.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+```
+TRAINING_IMAGES_DIRECTORY: Path to the directory containing training images.
+TRAINING_ANNOTATIONS_PATH: Path to the annotation file of the training set.
+VALIDATION_IMAGES_DIRECTORY: Path to the directory containing validation images.
+VALIDATION_ANNOTATIONS_PATH: Path to the annotation file of the validation set.
+OUTPUT_DIRECTORY: Path to the output directory, where the model will be saved.
+BATCH_SIZE: Size of training and validation batches. Default 16.
+EPOCHS: Number of epochs. Default 50.
+``` 
+We also provide a [Jupyter notebook](src/evaluation/evaluation.ipynb) for model evaluation, so one can quickly check quantitative and qualitative metrics as well as export them to a PDF file.
```
Comment (brunodrugowick) [inline README.md:79]:
```suggestion

We also provide a [Jupyter notebook](src/evaluation/evaluation.ipynb) for model evaluation, so one can quickly check quantitative and qualitative metrics as well as export them to a PDF file.
```

**Review by brunodrugowick — COMMENTED** [review]
This first review is just a bunch of nit-picks. You can ignore most of them if you prefer.

File: docker/build-container.sh [hunk 18]
```diff
@@ -3,12 +3,37 @@
...
-        docker rm cv-hardware-parts
-	docker rmi ia-bigdata
-    fi;
+while [[ $# -gt 0 ]]; do
+    case $1 in 
+    --clean-container | --clean-all)
+        docker rm ai-palatal-rugoscopy-env > /dev/null 2>&1
+        ;;
+    --clean-image | --clean-all)
+        docker rmi ai-palatal-rugoscopy > /dev/null 2>&1
```
Comment (brunodrugowick) [inline docker/build-container.sh]: `ai-palatal-rugoscopy` and `ai-palatal-rugoscopy-env` are used several times in the script. Make them constants.

File: docker/build-container.sh [hunk 19]
```diff
@@ -3,12 +3,37 @@
 export UID=$(id -u)
 export GID=$(id -g)
 
-for arg in $@; do
-    if [[ $arg == clean ]]; then
-        docker rm cv-hardware-parts
-	docker rmi ia-bigdata
-    fi;
+while [[ $# -gt 0 ]]; do
+    case $1 in
```
Comment (brunodrugowick) [inline docker/build-container.sh:10]: We could have a `--help`  here, then we might not need details in the README (maybe).

But then it's better to only set up variables in the `case` block instead of actually taking actions (leaving actions for later down the script), then the help calls a help function and exits immediately, so there's no effect in running the script with the help.

File: src/training/train.py [hunk 20]
```diff
@@ -1,16 +1,9 @@
-# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
-# Date: 2023-04-01                                                                                          #
-# Author: Joao Herrera                                                                                      #
```
Comment (brunodrugowick) [inline src/training/train.py:3]: You can leave the authorship, I believe.

File: src/training/train.py
Comment (brunodrugowick) [inline src/training/train.py:1]: Sorry, I don't have much to say about all this 😆

**Review by brunodrugowick — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (joaoherrera) [reply README.md:46]: Indeed :laughing:. It is all about avoiding ugly blocks of code/text. We can establish a pattern if you wish.

Comment (brunodrugowick) [reply README.md:46]: Most of what I'm suggesting here is that both titles and code blocks are separated by empty lines (above and below) from the rest of the text.

Comment (brunodrugowick) [reply README.md:46]: The only exception could be (but I honestly prefer it to be the same) code blocks inside bullet points.

File: docker/build-container.sh
Comment (brunodrugowick) [inline docker/build-container.sh:1]: Make sure this is executable `chmod +x build-container.sh`.

File: docker/build-container.sh [hunk 21]
```diff
@@ -3,12 +3,37 @@
...
-docker run -it --gpus all -v $(dirname `pwd`):/home/$USER/workspace/cv-hardware-parts -v $HOME/SMACS-0723:/home/$USER/server --name cv-hardware-parts ia-bigdata
+# Check whether path to data folder is provided by the user, otherwise no data will be available inside the container.
+if [[ -z $data_path ]]; then
+    echo "ERROR: Data folder not specified"
+    exit 1
+fi
+
+docker build \
+    --build-arg USER=$USER \
+    --build-arg PW=88452452 \
```
Comment (brunodrugowick) [inline docker/build-container.sh]: Is this a password or something?

File: docker/build-container.sh (hunk 19 above)
Comment (joaoherrera) [reply docker/build-container.sh:10]: I agree. But:

1- I thought it worth it having the script details in the README file, since different types of researchers (especially those who might not be well-versed with command line) may access this repository. It is a minor thing, but having such information in a README file could facilitate their understanding somehow. On the other hand, we need to be careful about providing a lot of information in the home page. Just to think about it.

2- I would move this task to another issue (e.g. refactor and clean the code).

File: docker/build-container.sh (hunk 21 above)
Comment (joaoherrera) [reply docker/build-container.sh]: Yes. I wasn't so inspired when creating this one :laughing:. 
I'm not sure if we really need this, since a default password is already being defined in the Dockerfile.

Comment (joaoherrera) [reply docker/build-container.sh]: Just tested without passing the password and it works. I will remove it from the shell script.

File: docker/build-container.sh (hunk 19 above)
Comment (brunodrugowick) [reply docker/build-container.sh:10]: Works for me! 😄

**Review by brunodrugowick — APPROVED** [review]
